spinalcordtoolbox/image.py

```diff
@@ -279,7 +282,22 @@ def __init__(self, param=None, hdr=None, orientation=None, absolutepath=None, di
 
         # load an image from file
         if isinstance(param, str):
-            self.loadFromPath(param, verbose)
+            try:
+                self.loadFromPath(param, mmap, verbose)
+            except OSError as e:
+                if str(e) == "[Errno 24] Too many open files":
+                    try:
+                        # Only load images into memory if there are no more available file descriptors
+                        self.loadFromPath(param, False, verbose)
```

Comment:
> Edit: Or is the constrained resource actually memory mappings, even though it's reported as too many open files? Otherwise how does this fix work?

As far as I understand, ever since https://github.com/spinalcordtoolbox/spinalcordtoolbox/pull/3718, we no longer keep any open file descriptors for `Image` objects (when loaded as `ndarray`s. In other words, it's only when the data array is loaded as a memory mapped array that we get an open file descriptor.

So, my understanding was that, by forcing a data array to be loaded as an `ndarray`, we ensure that we can have 300+ `Image` objects and 0 open file descriptors.

> For another thing, running out of file descriptors is a little bit like running out of memory: unless we actively free up some resources, everything is likely to start failing. For example, opening a file to write the results of whatever computation we're doing will likely fail. Also, import statements for not-already-imported modules would also fail, etc.

That was exactly my thinking in the extended commit message of https://github.com/spinalcordtoolbox/spinalcordtoolbox/commit/cf82f856cd30cc5013bdcb39eb5275549dc60825, too. 

I'll revert this solution and stick with the others.

Reply:
Oh I hadn't seen your lovely detailed commit messages! Sorry, I usually look for those but forgot in this PR.